File: python/src/facebook_engagement_mcp/outcomes/writes.py
"""The two write tools.

Both take ``page_id`` as **required**, and that is the fix for a full day lost
(T-26, 2026-09-08). They used to accept it optionally and fall back to the
startup client, which holds the USER token — so a caller supplying only a comment
id got Meta refusing the write by naming ``publish_actions``, a permission
removed in 2018. That message is literally true and entirely misleading: the
fault is the identity, not the permission, and no App Review can grant a dead
scope. Refusing here costs one round trip and says the true thing.

There is no delete tool, deliberately. Admitting one requires a decision record,
not a flag.
"""

from __future__ import annotations

from typing import Any

from ..graph import MetaApiError, PagesClient

NO_PAGE_ID = (
    "A reply is published as the Page, so it needs a Page token. Supply `pageId` for the "
    "Page that owns this comment. Without it the call would go out as the user, which Meta "
    "refuses with a message about `publish_actions` — a permission removed in 2018."
)


async def run_respond_to_comment(
    client: PagesClient, *, comment_id: str, message: str,
    page_id: str | None = None, dry_run: bool = False,
) -> dict[str, Any]:
    text = message.strip()
    if not text:
        return {"error": "A reply needs a non-empty message."}
    if not page_id:
        return {"error": NO_PAGE_ID}

    # Checked AFTER the pageId guard, so a dry run can never report that a call
    # would work when it could not.
    if dry_run:
        # The shape of the effect, never the text itself — the caller wrote it and
        # echoing it back costs context for nothing.
        return {"dryRun": True,
                "intended": {"commentId": comment_id, "messageLength": len(text),
                             "visibility": "public"}}

    try:
        created = await client.reply(comment_id, text)
    except MetaApiError as error:
        return {"error": str(error), "code": error.code}
    return {"ok": True, "replyId": created.get("id"), "commentId": comment_id}


async def run_moderate_comment(
    client: PagesClient, *, comment_id: str, action: str,
    page_id: str | None = None, dry_run: bool = False,
) -> dict[str, Any]:
    if action not in ("hide", "unhide"):
        return {"error": "action must be 'hide' or 'unhide'."}
    if not page_id:
        return {"error": NO_PAGE_ID.replace("A reply is published as", "Hiding acts as")}

    if dry_run:
        return {"dryRun": True, "intended": {"commentId": comment_id, "action": action}}

    try:
        await client.set_hidden(comment_id, action == "hide")
    except MetaApiError as error:
        return {"error": str(error), "code": error.code}
    return {"ok": True, "commentId": comment_id, "action": action}

File: go/internal/outcomes/writes.go
package outcomes

import (
	"context"
	"errors"
	"strings"
)

// ModerationAction is what moderate_comment does to a comment.
type ModerationAction string

const (
	ActionHide   ModerationAction = "hide"
	ActionUnhide ModerationAction = "unhide"
)

// Actions is every accepted value, for schema generation and validation.
var Actions = []ModerationAction{ActionHide, ActionUnhide}

// RespondOptions is the reply tool's input.
type RespondOptions struct {
	CommentID string
	Message   string
	PageID    string
	DryRun    bool
}

// ModerateOptions is the moderation tool's input.
type ModerateOptions struct {
	CommentID string
	Action    ModerationAction
	PageID    string
	DryRun    bool
}

// missingPageIDMessage is the refusal both write tools give.
//
// Without a pageId the call would fall through to the startup client, which
// holds the USER token — and Meta refuses a user-token comment write by naming
// `publish_actions`, dead since 2018. Refusing here costs one round trip and
// says the true thing; letting it through costs a day. Observed 2026-09-08.
func missingPageIDMessage() string {
	return "A reply is published as the Page, so it needs a Page token. Supply `pageId` for the " +
		"Page that owns this comment. Without it the call would go out as the user, which Meta " +
		"refuses with a message about `publish_actions` — a permission removed in 2018."
}

// RunRespondToComment publishes a reply to a comment, as the Page.
func RunRespondToComment(ctx context.Context, deps Deps, opts RespondOptions) (map[string]any, error) {
	message := strings.TrimSpace(opts.Message)
	if message == "" {
		return nil, errors.New("a reply needs a non-empty message")
	}

	// The pageId guard runs BEFORE the dry run, so a dry run can never report
	// that a call would work when it could not. The tool schema also marks
	// pageId required; either layer satisfies the rule, and both are kept.
	if opts.PageID == "" {
		return nil, errors.New(missingPageIDMessage())
	}

	if opts.DryRun {
		// Report the SHAPE of the effect, never the text itself — the caller wrote
		// it and echoing it back costs context for nothing.
		return map[string]any{
			"dryRun": true,
			"intended": map[string]any{
				"commentId":     opts.CommentID,
				"messageLength": len([]rune(message)),
				"visibility":    "public",
			},
		}, nil
	}

	tasks, _ := deps.Tokens.TasksFor(ctx, opts.PageID)
	token, err := deps.Tokens.TokenFor(ctx, opts.PageID)
	if err != nil {
		return nil, err
	}
	client, err := deps.NewPagesClient(token)
	if err != nil {
		return nil, err
	}

	// No retry. A retried reply is a double post — the transport's Post path
	// enforces that, and nothing here may route around it.
	replyID, err := client.Reply(ctx, opts.CommentID, message)
	if err != nil {
		return nil, errors.New(ExplainGraphError(err, ErrorContext{
			Operation: "reply", PageID: opts.PageID, Tasks: tasks,
		}))
	}

	return map[string]any{"ok": true, "replyId": replyID, "commentId": opts.CommentID}, nil
}

// RunModerateComment hides or unhides a comment.
func RunModerateComment(ctx context.Context, deps Deps, opts ModerateOptions) (map[string]any, error) {
	if opts.Action != ActionHide && opts.Action != ActionUnhide {
		return nil, errors.New("action must be hide or unhide")
	}

	// Same reason as the reply tool, and checked before the dry run for the same
	// reason: a dry run must not report that a call would work when it could not.
	if opts.PageID == "" {
		return nil, errors.New("Hiding a comment acts as the Page, so it needs a Page token. " +
			"Supply `pageId` for the Page that owns this comment. Without it the call would go " +
			"out as the user, which Meta refuses with a message about `publish_actions` — a " +
			"permission removed in 2018.")
	}

	if opts.DryRun {
		return map[string]any{
			"dryRun": true,
			"intended": map[string]any{
				"commentId": opts.CommentID,
				"action":    string(opts.Action),
			},
		}, nil
	}

	hidden := opts.Action == ActionHide

	tasks, _ := deps.Tokens.TasksFor(ctx, opts.PageID)
	token, err := deps.Tokens.TokenFor(ctx, opts.PageID)
	if err != nil {
		return nil, err
	}
	client, err := deps.NewPagesClient(token)
	if err != nil {
		return nil, err
	}

	if err := client.SetHidden(ctx, opts.CommentID, hidden); err != nil {
		return nil, errors.New(ExplainGraphError(err, ErrorContext{
			Operation: "moderate", PageID: opts.PageID, Tasks: tasks,
		}))
	}

	return map[string]any{
		"ok":        true,
		"commentId": opts.CommentID,
		"action":    string(opts.Action),
		"hidden":    hidden,
	}, nil
}
